### PR TITLE
Some enhancements in support of interfaces in Pulse

### PR DESCRIPTION
**Pull request by aseemr**
The PR adds the following:
 
* When interleaving, use the extension parser peek functionality to determine the name of the extension definition. This allows matching a val in the interface to the extension definition in .fst.
* Adds support for expected type in DSL checking. When there is a val present for the extension definition, we need to check that the type of the definition is consistent with the val. We now pass an optional expected type to the DSL type checker and expect the proof that the returned sigelt has the expected type, if any.

### CODE REVIEW

File: src/tactics/FStar.Tactics.Hooks.fst [hunk 1]
```diff
@@ -831,27 +837,67 @@ let splice (env:Env.env) (is_typed:bool) (lids:list Ident.lident) (tau:term) (rn
...
-                sigmeta_already_checked = checked; }
-          }
-        )
-        in
-        gs, sigelts
+        //
+        // See if there is a val for the lid
+        //
+        if List.length lids > 1
+        then let s = lids |> List.map Ident.string_of_lid |> BU.concat_l ", " in
```
Comment (mtzguido) [inline src/tactics/FStar.Tactics.Hooks.fst:844]: `s` unused? :)

File: src/tactics/FStar.Tactics.Hooks.fst [hunk 2]
```diff
@@ -831,27 +837,67 @@ let splice (env:Env.env) (is_typed:bool) (lids:list Ident.lident) (tau:term) (rn
...
+          //
+          // The arguments to run_tactic_on_ps here are in sync with ulib/FStar.Tactics.dsl_tac_t
+          //
+          let gs, (sig_blobs_before, sig_blob, sig_blobs_after) = run_tactic_on_ps tau.pos tau.pos false
+            (e_tuple2 RE.e_env (e_option RE.e_term))
+            ({env with gamma=[]}, val_t)
+            (e_tuple3
+              (e_list (e_tuple3 e_bool RE.e_sigelt e_blob))
+              (e_tuple3 e_bool RE.e_sigelt e_blob)
+              (e_list (e_tuple3 e_bool RE.e_sigelt e_blob)))
```
Comment (mtzguido) [inline src/tactics/FStar.Tactics.Hooks.fst:882]: Not a big deal but maybe using `Typeclasses.solve` can fill in these embeddings.

**Review by mtzguido — APPROVED** [review]
Looks good! I like the new interface.

File: src/tactics/FStar.Tactics.Hooks.fst (hunk 1 above)
Comment (aseemr) [reply src/tactics/FStar.Tactics.Hooks.fst:844]: Ah yes, thanks! At some point I was adding this in the error message but then I switched to the show method.

**Comment by aseemr** [comment]
I have a local everest green, merging.

**Pull request merged**